File: js/closure.js
var a = 1;
function foo()
{
	var b = 2;
	g = 3;
	var f = function() 
	{
		var c = 4;
		return a+b+c;
	};
	f();
}

function foreach(list , func)
{
    while(list)
    {
        func(list.val)
        list=list.next;
    }
}

var list = null;
for(var i=0;i<4;i++)
{
    list = {var: i, next: list}
}

var i=0;
// the inner anonymous function can access the outter variable i
foreach(list, function(n){ console.log("node(" + i + ")=" + n); i++;});

function extent()
{
    var n = 0;
    return function(){
        n++;
        console.log("n="+n);
    }
}

f = extent();
f(); //n=1
f(); //n=2


//so, closure is a close environment structure that contain local variable 

function extend()
{
    return {
        val: 0,
        call: function() {
            this.val ++;
            console.log("val="+this.val);
        }
        }
}

f = extend();
f.call(); //val = 1
f.call(); //val = 2

File: lua/closure.lua
--[[
a closure is a function plus all it needs to access its upvalues correctly. 

Closures provide a valuable tool in many contexts. 
As we have seen, they are useful as arguments to higher-order functions such as sort. 

Closures are valuable for functions that build other functions too, like the newCounter example;

Closures are useful for callback functions too


--]]

function newCounter ()
   local i = 0
   return function ()   -- anonymous function
       i = i + 1
       return i
    end
end
    
c1 = newCounter()
print(c1())  --> 1
print(c1())  --> 2

c2 = newCounter()
print(c2())  --> 1
print(c1())  --> 3
print(c2())  --> 2

--redefine math.sin with new version, keep the old version in a private variable; the only way to access it is through the new version. 
do
   local oldSin = math.sin
   local k = math.pi/180
   math.sin = function (x)
       return oldSin(x*k)
   end
end
